src/protocol/prss.rs

```diff
@@ -151,24 +133,69 @@ impl RngCore for PrssRng {
 impl CryptoRng for PrssRng {}
 
 /// A single participant in the protocol.
-/// This holds multiple streams of correlated (pseudo)randomness, indexed by `I`.
-pub struct Participant<I: SpaceIndex> {
-    // This would be `[PrssSpace; I::MAX]` with `feature(generic_const_exprs)`,
-    // but that is still in Nightly.
-    spaces: Vec<PrssSpace>,
-    _marker: PhantomData<I>,
+/// This holds multiple streams of correlated (pseudo)randomness.
+pub struct Participant {
+    inner: Mutex<ParticipantInner>,
 }
 
-impl<I: SpaceIndex> Debug for Participant<I> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        write!(f, "Participant[0..{0}]", I::MAX)
+impl Participant {
+    /// Construct a new, unconfigured participant.  This can be configured by
+    /// providing public keys for the left and right participants to `setup()`.
+    pub fn prepare<R: RngCore + CryptoRng>(r: &mut R) -> ParticipantSetup {
+        ParticipantSetup {
+            left: KeyExchange::new(r),
+            right: KeyExchange::new(r),
+        }
+    }
+
+    /// Get the identified PRSS instance.
+    ///
+    /// # Panics
+    /// When used incorrectly.  For instance, if you ask for an RNG and then ask
+    /// for a PRSS using the same key.
+    pub fn prss(&self, key: impl AsRef<str>) -> &PrssSpace {
```

Comment:
wondering if we should wrap `str` into an enum with a member for debug builds (just a string) and maybe an integer for release. Would this API allow non-string keys without additional overhead?

Reply:
Yes, I think that is the next step, where we work out how to take the unique ID and turn it into a flat number (for which copying should be very cheap) for release builds.